File: practice/ruby/merge_sort.rb
def merge_sort array
	return array if array.length < 2

	middle = array.length / 2
	left = array[0...middle]
	right = array[middle..-1]

	stitch(merge_sort(left), merge_sort(right))
end

def stitch left, right
	results = []

	while left.length > 0 && right.length > 0
		if left[0] <= right[0]
			results << left.shift
		else
			results << right.shift
		end
	end

	results + left + right
end

array_to_sort = [3,1,3,6,8,5,3,4,8,9,8,6,3,46,7,2]

puts merge_sort(array_to_sort)

File: practice/mergeSort.js
const mergeSort = array => {
	if (array.length < 2) {
		return array;
	}

	let middle = Math.floor(array.length / 2);
	let left = array.slice(0, middle);
	let right = array.slice(middle, array.length);

	return stitch(mergeSort(left), mergeSort(right));
}

const stitch = (left, right) => {
	const results = []
	while (left.length && right.length) { // loop will continue until both lengths are 0
		if (left[0] <= right[0]) {
			results.push(left.shift());
		} else {
			results.push(right.shift());
		}

	}

	while (left.length) {
		results.push(left.shift());
	}
	while (right.length){
		results.push(right.shift());
	}

	return results;
};


let a = mergeSort([2,3,6,4,7,2,4,8,4,3,8,9,4,2,4,2])
console.log(a);
